### ISSUE TITLE
Provide tools for GDPR compliance

### ISSUE THREAD

**Issue by Shelob9** [CalderaWP/Caldera-Forms#2108]
# This is the overview/ roundup issue see [this page](https://calderaforms.com/gdpr) 

Given the amount of data collected by Caldera Forms, we must provide the right tools for GPDR compliance by the 28th of May 2018. I think this is a good thing as our users are sitting on a pile of private data and a lot of times, deleting it is great solution to avoid privacy concerns and trim database size.

Everything we do has to be opt-in by site admin. I think most sites should enable this compliance, but I don't think we should make that call for them and some may have other strategies for compliance.

We will use the standard proposed here: https://github.com/GDPRWP/standard

What we need is the following:

- New method to get all entries where user ID is a given ID.
- New method to get all entires where given email is the value of a given field.
- New method to delete entires matching those results.

- Admin-facing user interface screen for GPDR (maybe we call this "User Privacy")
-- Setting to enable discovery and request deletion
-- Setting to restrict look up to logged in users. This is the default, we can't just let anyone enter email addresses into the site 
-- Setting for 

- New form processor for GPDR
-- Search by email - return results -- return results with delete option
-- Search by current logged in user -- return results with delete option

- Template for a GPDR Form
-- Look up fields
-- Privacy policy display field

**Comment by UniversalDesignz** [comment]
Hi Josh,

You guys are doing an awesome job. I am looking forward to the new updates with GPDR compliance implemented in Caldera forms? It would be nice to have an option that would allow to choose auto deletion after scheduled day(s) | month(s) | Year(s) or indefinite time.

Thank you.
Kindest Regards
Jay

**Comment by Shelob9** [comment]
Update on our work on GDPR compliance:

We need two basic things:
1. A query builder tool that can select and delete the kind of data needed to comply with GDPR.
2. A system with an interface and stuff for configuring, managing and handling these requests and the privacy statement and...

🌋[Item one is now ready to use.](https://github.com/CalderaWP/caldera-forms-query). It is a composer package, not a plugin. It is a developer tool that is also going to be useful for improving entry viewing, editing, display and exporting to CSV.

🌋I met with Kevin from WP-GDPR today from [WP GDPR](https://wp-gdpr.eu). [It looks like EDD](https://github.com/easydigitaldownloads/easy-digital-downloads/issues/6216#issuecomment-378780795) is going to create an add-on for that plugin, and they sent me their CF7 add-on to look at. I think that either an add-on for WP-GDPR or bundled features in the plugin that just work when WP-GDPR is active is a good way to go.

**Comment by Shelob9** [comment]
Ok, core now has or will have what we need. My prior concern was that the milestone on GDPR tickets was 5.0, which makes Gutenberg, not GDPR deadline the blocker for release of these features. I'm willing to be bet that if these features are ready and Gutenberg isn't ready, a 4.9.x release will happen.

https://make.wordpress.org/core/2018/03/28/roadmap-tools-for-gdpr-compliance/

https://core.trac.wordpress.org/changeset/42889/trunk/src/wp-admin/includes/ajax-actions.php

So I'm going to work on a UI for Caldera Forms to identify fields as personally identifying #2426 and consent fields #2428 and then once core is ready, integrate the query builder with the core filters as needed for export/delete/etc of data #2429

**Comment by UniversalDesignz** [comment]
Hi  Josh,

I hope you have been well. Firstly thank you to you and Caldera team for all the updates and support. 
For PII will we get an auto email request sent to our email (inbox); regards to the visitor that has requested and which field needs to be removed or amended (this will help if we hold any offline data as well)?

I look forward to your response.

Kindest Regards
Jay

**Comment by scrobbleme** [comment]
I want to add one thing to your list:

**Global setting for activating an "acceptance checkbox" on all forms**

We have to add such a checkbox to every form, so it would make sense to have a global option for this:

- This should always be required
- Always prior the submit button
- Configurable text with the option to link to WordPress' defined privacy page

**Comment by Shelob9** [comment]
@scrobbleme We are adding a consent field (#2428) and a privacy magic tag if you want to make your own (#2520). Adding a field to every form globally is something we can consider for a future request. IT would be super complex to automatically put it in the right place in the form layout, especially with multi-page forms and conditional logic on submit buttons.

**Comment by PenniKabo** [comment]
From the above it looks like the consent tick box will be optional. I'd vote for it to be this way. Consent is only one legal basis, personally I'm using contract and/or legitimate interest and so are my clients. For this reason, we have a statement instead of an action.

**Comment by Shelob9** [comment]
@PenniKabo Yes, it will be something you have to add, either using the consent field or making your own.

**Comment by Shelob9** [comment]
🌋🌋 GDPR Update:
TL;DR-  Everything is basically done, but not assembled. Release candidate out tomorrow-ish.

Needs done before 1.7 can be released:
👀 Adding Caldera Forms data to WordPress' data exports and erasing data when requested: Done, not merged. I will complete testing tomorrow #2507
🌋 New privacy and data page: included in #2507
👀 Consent field: done. I will test and merge tomorrow
✅ Add suggested privacy policy copy: Done and merged #2497
✅ Option to delete all entries: Done and merged #2005
❌ Privacy page magic tag (for making your own consent field/ using HTML fields for privacy statements) : Can be done once consent field is done #2520

⏰⏰⏰⏰⏰ I am aware the deadline is approaching for GDPR compliance and am working as fast as I can, but must get this right without creating new technical debt and also I have to sleep and eat and such. 1.7 should be out this week. I expect a 1.7.1 with bug fixes to follow soon after.

**Comment by cklosowski** [comment]
| and also I have to sleep and eat and such. 1.7 should be out this week

Keep up the good work. As a fellow dev in the thick of their, I appreciate the effort and status update.

**Comment by Shelob9** [comment]
🌋🌋 GDPR Update:
All GDPR features for Caldera Forms are merged to develop branch.

✅  Consent fields and privacy page magic tag: https://calderaforms.com/doc/consent-field/
✅ Option to delete all entries of a from: https://calderaforms.com/doc/delete-saved-entries-caldera-form/
✅  Integration with core exporters and erasers + shiny new privacy page: https://calderaforms.com/doc/setting-caldera-forms-gdpr-data-requests/

**Comment by scrobbleme** [comment]
@Shelob9 1.7.0 beta 1 looks great and seems to be working so far.

A litte bit confusing, but nothing for the first release:

- The label for the consent field should be handled like normal (checkbox) labels, then there should be another field for the description and a smaller field for linking to the policy page (see my screenshot)
- The privacy configuration page is confusing. I thought as soon as I mark fields as PII, the data will be contained within wordpress privacy export. It is not obvious that it is necessary to enable each form.
- PII might be select by default 


![image](https://user-images.githubusercontent.com/504835/40528541-a05abd20-5ff1-11e8-89aa-234bacbfec79.png)

**Comment by UniversalDesignz** [comment]
Hi Josh @Shelob9 

**Consent Field > [ Linked Text ]:**
1. The text need to be smaller.
2. The whole sentence shouldn't be linked just the word or to that needs to be linked.
3. The default link styling of link are getting discarded.

**Control/Manage Delete/Amend Saved Entries?**
**1. Caldera> Forms Privacy I selected the form and enable the GDPR with the correct fields I needed (also clicked on PII in all the fields under the form), But when a test contact form was sent. The [ Add Data Export Request ] shows blank.**
2. It would be nice to have a scheduled deletion and disable data entries option.

**Comment by Shelob9** [comment]
@scrobbleme and @UniversalDesignz 5 imaginary internet points each for the awesome feedback, this helps a lot.

I agree with you both on the consent field that the linked text should be a separate setting. See: https://github.com/CalderaWP/Caldera-Forms/issues/2428#issuecomment-392093814

In reply to @UniversalDesignz comments above:
> 1. Caldera> Forms Privacy I selected the form and enable the GDPR with the correct fields I needed (also clicked on PII in all the fields under the form), But when a test contact form was sent. The [ Add Data Export Request ] shows blank.
Did you enable the data exporter for that form? That is a required step. See: https://calderaforms.com/doc/setting-caldera-forms-gdpr-data-requests/

> 2. It would be nice to have a scheduled deletion and disable data entries option.
I agree: #2550 

In reply to @scrobbleme comments above:
>The privacy configuration page is confusing. I thought as soon as I mark fields as PII, the data will be contained within wordpress privacy export. It is not obvious that it is necessary to enable each form.
I think having the PII checkbox in the form editor was a mistake and is why both you and @UniversalDesignz think that should be all is needed. I will remove it  #2551 All privacy settings will go in the privacy page is simpler. 

>PII might be select by default
Can I ask why? I think not be default as in a big form, a lot of fields might not be personally identifying. I could be backwards on this.

**Comment by Shelob9** [comment]
See changes to consent field here:

BTW If you tried this with beta 1, the settings are going to need for be updated. You will likely want to move what is in linked text to agreement.

Also @UniversalDesignz You mentioned concerns about style. I removed the inline style 588aa79f0f670b767478b12246a4078992b0326c and applied it with CSS stylesheet forms.css. 

I think it looks good and it can be changed with custom CSS. Do you have any specific changes you'd like to see about how it works now?

**Comment by UniversalDesignz** [comment]
@Shelob9 sorry about late response.

Did you enable the data exporter for that form? That is a required step. See: https://calderaforms.com/doc/setting-caldera-forms-gdpr-data-requests/
**Yes this was GDPR was been enabled. (Still not working...)**
![pii-settings](https://user-images.githubusercontent.com/15095585/40584644-573eb02c-619c-11e8-8322-cf81a12fcfdb.png)
**
I think having the PII checkbox in the form editor was a mistake and is why both you and @UniversalDesignz think that should be all is needed. I will remove it #2551 All privacy settings will go in the privacy page is simpler.
**Yes it's very confusing I thought that has to be ticked as to make the selected fields to function.
![read full policy link styling](https://user-images.githubusercontent.com/15095585/40585176-e38e0ff6-61a5-11e8-964a-5fcc69e7dc4a.png)


I hope the images help.

**Comment by godvsgodvs** [comment]
@Shelob9, Is there a more current beta available with these changes. I'm still getting version 1.7.0-b.1. If not, when should I check back to try them out.
Thanks,
Gary

**Comment by Shelob9** [comment]
@godvsgodvs  1.7.0 will be released this week. You can get the latest version by downloading the develop branch. git@github.com:CalderaWP/Caldera-Forms.git

@UniversalDesignz 
You showed a data export screen with no requests for data export. Did you create a request using an email address that you know an enabled form has PII for?

**Comment by Shelob9** [comment]
See here for a build of 1.7.0 https://github.com/CalderaWP/Caldera-Forms/pull/2563#issuecomment-393269578
